WebSocketServer: Expose the resource name the client used

Meriipu opened the pull request:
This adds a method to WebSocketServer to retrieve the resource name (the part after the domain name, including any queries) that the client connected with.

I did not expose it to clients both due to reduced code-complexity and the fact that the clients (at least non-html5 ones) should already have access to it implicitly. 

My use-cases for this are 
* Process a client immediately upon connection rather than wait for them to send a followup.
* Related to the above, possibly authentication of a client so that connections can be dropped immediately rather than possibly hanging open waiting for a bad client to authenticate itself in a followup. 

Very open to comments/feedback.

Meriipu committed:
Expose the resource name the client used to the websocket server

This information is exposed to the websocket server through the
client_connected-signal.

example.com/chat?id=10 gives the resource name "/chat?id=10"

Meriipu commented:
```
websocketserver.get_resource_name(1) 
```
(and also for non-existent ids) returns the empty string, as well as errors in the debugger:
```
E 0:00:02.849   get_peer_resource_name: Condition "!has_peer(p_peer_id)" is true. Returned: String()
  <C++ Source>  modules/websocket/wsl_server.cpp:288 @ get_peer_resource_name()
  <Stack Trace> CmdServerSetup.gd:45 @ _connected()
                CmdServerSetup.gd:77 @ _process()
```

I am not sure if it is desirable to handle this a different way. I am also not sure what happens if websockets are used in conjunction with high level multiplayer.

akien-mga commented on modules/websocket/wsl_server.cpp:
```diff
@@ -43,7 +43,7 @@ WSLServer::PendingPeer::PendingPeer() {
 }
 
-bool WSLServer::PendingPeer::_parse_request(const Vector<String> p_protocols) {
+bool WSLServer::PendingPeer::_parse_request(const Vector<String> p_protocols, String &resource_name) {
```
As it's passed as a reference, the parameter should be named `r_resource_name` (same in `do_handshake`).

Meriipu replied:
I take it the original variable `resource_name` should **not** get the `r_` up until and including the point where it appears as an argument in the first call?

akien-mga replied:
Yeah it's only in the function prototype, and thus in the function itself. But in the calling function where it's a local variable, it takes no prefix (same thing for `p_` read-only parameters).

Meriipu commented:
Does this need a proposal to be considered?

Faless commented:
> Does this need a proposal to be considered?

No, I think this is nice, I don't like too much that we keep the resource name for the whole life cycle.
I'd also drop `get_peer_resource_name` and just leave `get_peer(ID).get_resource_name()` in case.

Meriipu commented:
> > Does this need a proposal to be considered?
> 
> No, I think this is nice, I don't like too much that we keep the resource name for the whole life cycle.
>
The lifecycle of the peer? Or have I mistakenly stored it outside the peer/outside something that gets cleared when a peer is cleared?

> I'd also drop `get_peer_resource_name` and just leave `get_peer(ID).get_resource_name()` in case.
>
What distinguishes `get_peer_resource_name` from the other two server-methods `get_peer_port` or `get_peer_address`? That it is a higher-layer concept than address/port?
I think another part of the reason I put the method on the server rather than exposing the method on the peer itself is to avoid having clients also need to store the resource name when they connect (which increases the code-complexity a bit from what I could see), though I might be overlooking something silly or overestimating the extra complexity. I did not really like having to pass the extra reference argument through two calls on the server though.

I can see it being beneficial to more explicitly put it on the peer/connection itself though.

Meriipu commented:
I guess the client would be able to call it on a peer object (the server), but that it would not be useful (and that the server is the only peer the client is able to get with get_peer currently?).

Meriipu commented:
I am not sure if I understand what changes the review is requesting (or if I have misunderstood the increased complexity from the changes).

akien-mga commented:
> I am not sure if I understand what changes the review is requesting (or if I have misunderstood the increased complexity from the changes).

CC @Faless

Meriipu commented:
> > The lifecycle of the peer?
> 
> Yeah, that information is only relevant during connection, i.e. during the HTTP handshake, it's meaningless after that (from the protocol point of view), so it should not be stored anywhere after the handshake has completed. It might make sense to add it to the `client_connected` signal, but why should we keep it after that?

This makes sense, thanks. I did not think of signals when I thought of how to only temporarily expose a value.
Is it unproblematic to change the API by modifying existing signals (like `client_connected`) though? It would be added after the `peer_id` and `protocol` parameters from what I see.

> Yeah, you are right, it's the other way around, the method should not stay on the peer at all.
> Not `WebSocketPeer::get_resource_name` should exist in peer.
> 

On the bright side, if adjusting existing signals by adding another parameter is fine, then this will simplify the changes a lot since the `client_connected`-signal is emitted through a call to `_on_connect` in the very same function where the resourcename is obtained:
https://github.com/godotengine/godot/blob/77803c10c18299cd531e1649159a521fab418c97/modules/websocket/wsl_server.cpp#L200-L226

Faless commented:
> Is it unproblematic to change the API by modifying existing signals (like `client_connected`) though?

Yes, this breaks compatibility but it's fine to do in `4.0`.

Meriipu commented:
So more or less like this?
I am having trouble testing them because I only applied them to my 4.0 branch where the editor runs pretty bad for me.

Meriipu commented:
aaa forgot to update `ADD_SIGNAL`

Meriipu commented:
Still good to go, I think.

Meriipu commented:
Fixed a merge conflict in `modules/websocket/emws_client.cpp` since another commit also removed the superfluous semicolons there (there should now no changes to that file in this commit (though the previous changes there were just style-changes)).
Maybe it is better to squash and rebase instead of being lazy with the github web-ui though.

Faless reviewed (approved):
Approved in the network meeting, needs rebase.

Meriipu commented:
should be good now

although I have not been able to test it manually due to vulkan issues (the godot editor is detecting the wrong gpu/vulkan device and failing to start).

Faless commented:
Thanks!
